### PR TITLE
[RLlib] Dataset Reader/Writer for rllib

### PR DESCRIPTION
**Pull request by gjoliver**
## Why are these changes needed?

Use Dataset for offline RL I/O.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [*] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/trainer.py [hunk 1]
```diff
@@ -483,6 +483,8 @@
     #  - a path/URI to save to a custom output directory (e.g., "s3://bucket/")
     #  - a function that returns a rllib.offline.OutputWriter
     "output": None,
+    # Arguments accessible from the IOContext for configuring custom output
```
Comment (sven1977) [inline rllib/agents/trainer.py:486]: Very cool. Now we have this concept (defining `type`, or callable returning instance, or (str) classpath) for OutputWriter, Exploration, ReplayBuffer!
We should unify this way of configuring stuff more - in a separate PR, of course. :)

File: rllib/evaluation/worker_set.py [hunk 2]
```diff
@@ -397,6 +407,40 @@ def _from_existing(local_worker: RolloutWorker,
...
 
+    def _get_dataset_and_shards(self, config: TrainerConfigDict,
+                                num_workers: int, local_worker: bool)\
+            -> (ray.data.dataset.Dataset,
+                List[ray.data.dataset.Dataset]):
+        assert config["input"] == "dataset"
+        assert "input_config" in config, (
+            "Must specify input_config dict if using Dataset input.")
+
+        input_config = config["input_config"]
```
Comment (sven1977) [inline rllib/evaluation/worker_set.py:418]: Just a thought:
In a follow-up PR, we should unify these config keys. Maybe something like:

```
Dataset:
  input_config: {
    type: "dataset",
    path: [...]
  }

JSON File:
  input_config:
    type: "JSONReader",
    path: [...]  <- filename or path w/ JSON files in it

Sampler (default):
  input_config:
    type: "sampler",
    env(??):  # <- could replace the "env" config key
```

File: rllib/evaluation/worker_set.py [hunk 3]
```diff
@@ -397,6 +407,40 @@ def _from_existing(local_worker: RolloutWorker,
...
+        path = input_config["path"]
+        if type == "json":
+            dataset = data.read_json(path)
+        elif type == "parquet":
+            dataset = data.read_parquet(path)
+        else:
+            raise ValueError("Un-supported Ray dataset type: ", type)
+
+        if local_worker and num_workers == 0:
+            # Local worker will be responsible for sampling.
```
Comment (sven1977) [inline rllib/evaluation/worker_set.py:435]: Nit: Can we move these comments on top of the `if` and `else`?

File: rllib/evaluation/worker_set.py [hunk 4]
```diff
@@ -397,6 +407,40 @@ def _from_existing(local_worker: RolloutWorker,
...
+            # Dataset is the only shard we need.
+            return dataset, [dataset]
+        else:
+            # Each remote worker gets 1 shard.
+            # The first None shard is for the local worker, which
+            # shouldn't be doing rollout work anyways.
+            return dataset, [None] + dataset.split(num_workers, equal=True)
```
Comment (sven1977) [inline rllib/evaluation/worker_set.py:442]:
```suggestion
        # Remote workers responsible for sampling:
        else:
            # Each remote worker gets 1 shard.
            # The first None shard is for the local worker, which
            # shouldn't be doing rollout work anyways.
            return dataset, [None] + dataset.split(num_workers, equal=True)
```

File: rllib/offline/dataset_reader.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+import logging
```
Comment (sven1977) [inline rllib/offline/dataset_reader.py:1]: Nice!!

File: rllib/offline/dataset_reader.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        config = {
+            "input"="dataset",
+            "input_config"={
+                "type": "json",
+                "path": "/tmp/sample_batches/",
+            }
+        }
+
+    path may be a single data file, a directory, or anything
+        that ray.data.dataset recognizes.
```
Comment (sven1977) [inline rllib/offline/dataset_reader.py:28]:
```suggestion
    `path` may be a single data file, a directory, or anything
    that ray.data.dataset recognizes.
```

File: rllib/offline/dataset_writer.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
+import logging
```
Comment (sven1977) [inline rllib/offline/dataset_writer.py:1]: Nice! :)

**Review by sven1977 — COMMENTED** [review]
Awesome stuff! Could you fix the 2-3 nits I suggested (just formatting stuff). Thanks @gjoliver !

One conceptual question: Should we NOT soft-deprecate JSONReader/Writer? I feel like having these could still be nice for light-weight debugging purposes.
Maybe change the `@Deprecated` into a `logger.info("You are using JSONReader. It is recommended to use blabla, instead of JSONReader/Writer for better sharding support.")`

File: rllib/agents/trainer.py (hunk 1 above)
Comment (gjoliver) [reply rllib/agents/trainer.py:486]: Totally, let's discuss more. I am not clear how Exploration fits in this framework.
We should also unify InputReader with ReplayBuffer input I feel like.

File: rllib/evaluation/worker_set.py (hunk 2 above)
Comment (gjoliver) [reply rllib/evaluation/worker_set.py:418]: good idea. I added a big TODO near the input key.

File: rllib/evaluation/worker_set.py (hunk 3 above)
Comment (gjoliver) [reply rllib/evaluation/worker_set.py:435]: this particular comment is for the case when we don't have remote worker though. so I don't really have to split the ds.

Comment (gjoliver) [reply rllib/evaluation/worker_set.py:435]: oh, I see what you mean below, ok, done.

File: rllib/evaluation/worker_set.py [hunk 8]
```diff
@@ -397,6 +407,40 @@ def _from_existing(local_worker: RolloutWorker,
...
+            dataset = data.read_json(path)
+        elif type == "parquet":
+            dataset = data.read_parquet(path)
+        else:
+            raise ValueError("Un-supported Ray dataset type: ", type)
+
+        if local_worker and num_workers == 0:
+            # Local worker will be responsible for sampling.
+            # Dataset is the only shard we need.
+            return dataset, [dataset]
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/evaluation/worker_set.py:437]: done

File: rllib/evaluation/worker_set.py (hunk 4 above)
Comment (gjoliver) [reply rllib/evaluation/worker_set.py:442]: done

File: rllib/offline/dataset_reader.py (hunk 6 above)
Comment (gjoliver) [reply rllib/offline/dataset_reader.py:28]: done

**Comment by gjoliver** [comment]
> Awesome stuff! Could you fix the 2-3 nits I suggested (just formatting stuff). Thanks @gjoliver !
> 
> One conceptual question: Should we NOT soft-deprecate JSONReader/Writer? I feel like having these could still be nice for light-weight debugging purposes. Maybe change the `@Deprecated` into a `logger.info("You are using JSONReader. It is recommended to use blabla, instead of JSONReader/Writer for better sharding support.")`

sounds good.
added a couple of TODOs, plus print a message recommending folks to use DatasetR/W instead.
thanks for the suggestion.

**Review by sven1977 — APPROVED** [review]
Looks gooood! :)

**Pull request merged**